### PR TITLE
210 ownedpositionidsrequest

### PR DESCRIPTION
**Pull request by Valentine1898**

### CODE REVIEW

**Comment by Valentine1898** [comment]
I need to add the `POSITIONS` table to indexed-db and add saving the position data to` block-processor`, but it makes sense to do it after the refactoring in #357 is complete

**Comment by grod220** [comment]
`0.65.0-pre.1` published to npm with the new `compute_position_id()` available to use

**Comment by Valentine1898** [comment]
image
It's weird, broken tests that aren't related to this PR

File: packages/query/src/block-processor.ts [hunk 1]
```diff
@@ -29,6 +29,7 @@ import {
   TransactionInfo,
 } from '@buf/penumbra-zone_penumbra.bufbuild_es/penumbra/view/v1alpha1/view_pb';
 import { StateCommitment } from '@buf/penumbra-zone_penumbra.bufbuild_es/penumbra/crypto/tct/v1alpha1/tct_pb';
+import { computePositionId } from '@penumbra-zone/wasm-ts/src/dex';
```
Comment (grod220) [inline packages/query/src/block-processor.ts:32]: Let's add `export * from './src/dex';` to `packages/wasm/index.ts` and we can shorten the import like the others

File: packages/storage/src/indexed-db/index.ts [hunk 2]
```diff
@@ -340,4 +347,46 @@ export class IndexedDb implements IndexedDbInterface {
...
+      const positionId = PositionId.fromJson(positionCursor.value.id);
+      const position = Position.fromJson(positionCursor.value.position);
+
+      if (positionState && !positionState.equals(position.state)) {
+        continue;
+      }
+      if (tradingPair && !tradingPair.equals(position.phi?.pair)) {
+        continue;
+      }
+      ownedPositionIds.push(positionId);
```
[the remark this answers is not in the document]
Comment (Valentine1898) [reply packages/storage/src/indexed-db/index.ts:367]: Good idea, will also have to revise tests

**Comment by Valentine1898** [comment]
a lot of files changed just because of lint fix from #388

File: packages/router/src/grpc/view-protocol-server/index.ts [hunk 3]
```diff
@@ -27,7 +28,7 @@ import { walletId } from './wallet-id';
 import { witness } from './witness';
 import { witnessAndBuild } from './witness-and-build';
 
-export const viewImpl: Omit<Impl, 'ownedPositionIds'> = {
+export const viewImpl: Impl = {
```
Comment (grod220) [inline packages/router/src/grpc/view-protocol-server/index.ts:31]: giphy

File: packages/storage/src/indexed-db/index.ts [hunk 4]
```diff
@@ -349,4 +356,48 @@ export class IndexedDb implements IndexedDbInterface {
...
+
+      if (positionState && !positionState.equals(position.state)) {
+        continue;
+      }
+      if (tradingPair && !tradingPair.equals(position.phi?.pair)) {
+        continue;
+      }
+      yield positionId;
+    }
+  }
```
Comment (grod220) [inline packages/storage/src/indexed-db/index.ts:376]: Fields should be required here:


```typescript
  async *getOwnedPositionIds(
    positionState: PositionState,
    tradingPair: TradingPair,
  ): AsyncGenerator<PositionId, void> {
    for await (const positionCursor of this.db.transaction('POSITIONS').store) {
      const position = Position.fromJson(positionCursor.value.position);

      if (!positionState.equals(position.state) || !tradingPair.equals(position.phi?.pair)) {
        continue;
      }

      yield PositionId.fromJson(positionCursor.value.id);
    }
  }
```

File: packages/router/src/grpc/view-protocol-server/ownned-position-ids.ts [hunk 5]
```diff
@@ -0,0 +1,13 @@
+import type { Impl } from '.';
+import { servicesCtx } from '../../ctx';
+
+export const ownedPositionIds: Impl['ownedPositionIds'] = async function* (req, ctx) {
+  const services = ctx.values.get(servicesCtx);
+
+  const { indexedDb } = await services.getWalletServices();
+
+  const positionIds = indexedDb.getOwnedPositionIds(req.positionState, req.tradingPair);
```
Comment (grod220) [inline packages/router/src/grpc/view-protocol-server/ownned-position-ids.ts:9]: We should do parameter validation here:

```typescript
  if (!req.positionState) throw new Error('Missing position state in request');
  if (!req.tradingPair) throw new Error('Missing trading pair in request');

  const positionIds = indexedDb.getOwnedPositionIds(req.positionState, req.tradingPair);
  
```

Comment (Valentine1898) [reply packages/router/src/grpc/view-protocol-server/ownned-position-ids.ts:9]: But `positionState` and `tradingPair` should not be required. 
If they are required, the client will have to brute force all combinations of trading pairs and state to get all positions

Comment (Valentine1898) [reply packages/router/src/grpc/view-protocol-server/ownned-position-ids.ts:9]: https://github.com/penumbra-zone/penumbra/blob/a483d8e377ea6728f865d844f9adc4e7fe8f8827/crates/view/src/storage.rs#L1655 
In rust, they are also not required

Comment (grod220) [reply packages/router/src/grpc/view-protocol-server/ownned-position-ids.ts:9]: Ah, I misread the protos, thanks for correcting!

File: packages/storage/src/indexed-db/index.ts [hunk 6]
```diff
@@ -349,4 +356,48 @@ export class IndexedDb implements IndexedDbInterface {
     }
     return Promise.resolve(notesForVoting);
   }
+
+  async *getOwnedPositionIds(
+    positionState: PositionState | undefined,
+    tradingPair: TradingPair | undefined,
+  ): AsyncGenerator<PositionId, void> {
+    for await (const positionCursor of this.db.transaction('POSITIONS').store) {
+      const positionId = PositionId.fromJson(positionCursor.value.id);
```
Comment (grod220) [inline packages/storage/src/indexed-db/index.ts:365]: Think we can still drop this below so it doesn't deserialize unnecessarily:

```typescript
yield PositionId.fromJson(positionCursor.value.id);
```

**Review by grod220 — APPROVED** [review]
Nice job! Just one additional change I'd say.

File: packages/router/src/grpc/view-protocol-server/index.ts (hunk 3 above)
Comment (TalDerei) [reply packages/router/src/grpc/view-protocol-server/index.ts:31]: omit = null

**Pull request merged**